Wrap op params for [Gpu]DimShuffle.

notoraptor opened the pull request:
As a quick benchmark, running:
```
nosetests -xvs
 theano\tensor\tests\test_basic.py:test_batched_dot
 theano\tensor\tests\test_basic.py:test_batched_dot_not_contiguous
 theano\tensor\tests\test_basic.py:BatchedDotTester
```
With `master` branch (after `theano-cache purge`), 70.428s and 11 compilations for CPU DimShuffle.

With this branch (after `theano-cache purge`), 62.231s and 1 compilation for CPU DimShuffle.

@abergeron @lamblin

notoraptor commented:
I forgot to update GpuDimShuffle.

notoraptor commented:
Updated. I think GpuDimShuffle should be rewritten in a separate PR, as it is not trivial !

abergeron commented on theano/tensor/elemwise.py:
```diff
@@ -265,98 +278,113 @@ def c_code(self, node, name, inp, out, sub):
+            PyErr_SetString(PyExc_RuntimeError, "DimShuffle: param input_broadcastable must be C-contiguous.");
+            %(just_fail)s
+        }
+        if (!PyArray_IS_F_CONTIGUOUS(%(params)s->_new_order)) {
```
You check for F_CONTIGUOUS here.

notoraptor replied:
Oh, what a mistake. Sorry.

abergeron commented:
With the scope of the changes to DimShuffle, you might even consider moving to COp.

notoraptor commented:
@abergeron Done. Some remarks:
- as GpuDimShuffle inherits from DimShuffle which is now a COp, I added an empty C file for GpuDimShuffle to prevent file loading error. I have not rewritten the GPU op in that PR.
- from memory, there may be some pickling issues with old objects when we convert an Op to a COp, so I added a `__setstate__()` implementation to prevent it.

lamblin commented on theano/gpuarray/c_code/dimshuffle.c:
```diff
@@ -0,0 +1 @@
+#section support_code
```
I thought it would not be necessary, since `GpuDimShuffle` inherits from `HideC`.
@abergeron did I miss something?
I don't really like the idea of an almost-empty file.

notoraptor replied:
@lamblin In fact, GpuDimShuffle inherits from DimShuffle which is now a COp. But DimShuffle has an `__init__()` in which there is a call to `COp.__init__(self, [filenames], function_name)`. Then this `__init__()` is inherited into GpuDimShuffle, even if C functions are hidden.

I don't know how to bypass that. Maybe should I just rewrite GpuDimShuffle too in that PR as soon as possible ? (Monday, I think !)

abergeron replied:
GpuDimShuffle inherits from HideC and DimShuffle.  HideC will ensure that we don't try to reuse the C code from DimShuffle, so you could probably just add an empty `__init__` to GpuDimShuffle.

If you intend on rewriting GpuDimShuffle soon, even if not in this PR, I would say the current fix is fine.

notoraptor replied:
In fact, some important attributes like `self.suffle`, `self.drop`, `self.new_order` are created and checked  into DimShuffle `__init__()`, so this method is necessary for GpuDimShuffle too.

notoraptor replied:
I may rewrite GpuDimShuffle today.



NB: small remark: One of the problems when we rewrite ops is that we now need to allocate small arrays (`int* shape = (int*)malloc(params->size * sizeof(int))`), where we previously just hardcoded it (`int shape[%(size)s]`). It's quite regrettable, but I don't find a way to do it otherwise.

abergeron replied:
We could also go with a "max case" array of like 32 and check that the input array doesn't have more dimensions than that.

notoraptor replied:
Indeed, it's generally about array of strides or dimensions. But are we sure that users always work with at most 32 dimensions ^^ ? I think we can't do that assumption, unfortunately.

abergeron replied:
NumPy has that limit.  libgpuarray doesn't currently, but I'm thinking about adding it in, since it would simplify some things.

lamblin replied:
What about `alloca`?

notoraptor replied:
Not sure that alloca is portable across systems and compilers:
- http://man7.org/linux/man-pages/man3/alloca.3.html#NOTES
- https://github.com/Theano/Theano/issues/6094

notoraptor commented:
Updated (with malloc()). GpuDimShuffle rewritten.

notoraptor commented:
Jenkins fails, but I don't know why: http://earlgrey.iro.umontreal.ca:8080/job/Theano_PR/461/#showFailuresLink

jenkins test this

notoraptor commented:
Jenkins succeeded.

lamblin commented on theano/gpuarray/elemwise.py:
```diff
@@ -563,7 +504,7 @@ def __init__(self, scalar_op, axis=None,
         self.reduce_mask = reduce_mask
 
-        # used to make sure that calls to scalar op
+        # used to make sure that callfs to scalar op
```
Probably a typo here

notoraptor replied:
Indeed, I'll fix it!

notoraptor commented:
Typo fixed @lamblin .

abergeron commented on theano/gpuarray/c_code/dimshuffle.c:
```diff
@@ -0,0 +1,81 @@
+    }
+    tmp = pygpu_transpose(input, transposition);
+    if (!tmp) {
+        PyErr_SetString(PyExc_RuntimeError, "GpuDimShuffle: unable to transpose input.");
```
There is already a python exception here.

notoraptor replied:
Ah ok, I see.

abergeron commented on theano/gpuarray/c_code/dimshuffle.c:
```diff
@@ -0,0 +1,81 @@
+        tmp = pygpu_copy(*out, GA_ANY_ORDER);
+        Py_DECREF(*out);
+        if (!tmp) {
+            PyErr_SetString(PyExc_RuntimeError, "GpuDimShuffle: unable to copy output.");
```
Masking exception

abergeron commented on theano/gpuarray/c_code/dimshuffle.c:
```diff
@@ -0,0 +1,81 @@
+    /** End shuffle. **/
+
+    if (*out == NULL) {
+        PyErr_SetString(PyExc_RuntimeError, "GpuDimShuffle: unable to reshape output.");
```
Masking exception.

abergeron commented on theano/gpuarray/c_code/dimshuffle.c:
```diff
@@ -0,0 +1,81 @@
+        return 1;
+    }
+    for (npy_intp i = 0; i < nd_in; ++i) {
+        transposition[i] = ((npy_int64*) PyArray_DATA(params->transposition))[i];
```
You could set the type of transposition to uintt32 and use it directly to avoid a malloc and copy.

notoraptor replied:
`pygpu_transpose` expects transposition to be an array of `unsigned int`, but we can't be sure that `int` it always 32 bits in all machines and computers.

NB: I will make a commit to suggest a solution in next hours.

abergeron replied:
We know that this is the case on all platforms that we care about and we can always add an assert so that we get a failure rather than wrong results on exotic platforms.

notoraptor commented:
@abergeron I was working on a way to get sizeof a machine-dependent C type, so I added the work here. What do you think ? If it's unsafe or too complicated, I can remove it and just put an assert as you suggest.

notoraptor commented:
Code ultimately updated and simplified (transposition is simply an array of `npy_uint32`) @abergeron .

abergeron commented on theano/tensor/elemwise.py:
```diff
@@ -185,6 +206,13 @@ def __init__(self, input_broadcastable, new_order, inplace=True):
         if self.inplace:
             self.view_map = {0: [0]}
```
I just thought about this, but don't we enable views all the time.  Usually inplace is for destroy_map.

@lamblin ?

notoraptor replied:
Should I change `view_map` with `destroy_map` ?

abergeron replied:
No, because this doesn't actually destroy anything.  I am saying that we can probably remove the inplace property and act as if it was always True.
